openverse_catalog/dags/common/popularity/sql.py

```diff
@@ -308,19 +308,26 @@ def create_audioset_view_query():
         f"""
         CREATE VIEW public.{AUDIOSET_VIEW_NAME}
         AS
-        SELECT DISTINCT
-            (audio_set ->> 'foreign_identifier')  :: varchar(1000) as foreign_identifier,
-            (audio_set ->> 'title')               :: varchar(2000) as title,
-            (audio_set ->> 'foreign_landing_url') :: varchar(1000) as foreign_landing_url,
-            (audio_set ->> 'creator')             :: varchar(2000) as creator,
-            (audio_set ->> 'creator_url')         :: varchar(2000) as creator_url,
-            (audio_set ->> 'url')                 :: varchar(1000) as url,
-            (audio_set ->> 'filesize')            :: integer       as filesize,
-            (audio_set ->> 'filetype')            :: varchar(80)   as filetype,
-            (audio_set ->> 'thumbnail')           :: varchar(1000) as thumbnail,
+          -- DISTINCT clause exists to ensure that only one record is present for a given
+          -- foreign identifier/provider pair. This exists as a hard constraint in the API table
+          -- downstream, so we must enforce it here. The audio_set data is chosen by which audio
+          -- record was most recently updated (see the final section of the ORDER BY clause
+          -- below). More info here:
+          -- https://github.com/WordPress/openverse-catalog/issues/658
+          SELECT DISTINCT ON (audio_set ->> 'foreign_identifier', provider)
+            ((audio_set ->> 'foreign_identifier'::text))::character varying(1000) AS foreign_identifier,
+            ((audio_set ->> 'title'::text))::character varying(2000) AS title,
+            ((audio_set ->> 'foreign_landing_url'::text))::character varying(1000) AS foreign_landing_url,
+            ((audio_set ->> 'creator'::text))::character varying(2000) AS creator,
+            ((audio_set ->> 'creator_url'::text))::character varying(2000) AS creator_url,
+            ((audio_set ->> 'url'::text))::character varying(1000) AS url,
+            ((audio_set ->> 'filesize'::text))::integer AS filesize,
+            ((audio_set ->> 'filetype'::text))::character varying(80) AS filetype,
+            ((audio_set ->> 'thumbnail'::text))::character varying(1000) AS thumbnail,
             provider
-        FROM public.{AUDIO_VIEW_NAME}
-        WHERE audio_set IS NOT NULL;
+          FROM public.{AUDIOSET_VIEW_NAME}
```

Comment:
This should be `FROM public.{AUDIO_VIEW_NAME}`, right?

Reply:
Oh goodness, good catch!